### PR TITLE
Loosen ParameterValue dtype contract via Array-API Protocol

### LINKED ISSUE #270: Loosen ParameterValue dtype/backend contract via Array-API Protocol
## Environment

- OS: macOS 15
- Python version: 3.13
- `flepimop2` version: `0.3.0.dev0` (current `main`)

## Reproducible example

```python
import jax, jax.numpy as jnp
from flepimop2.axis import ResolvedShape
from flepimop2.parameter.abc import ParameterValue

shape = ResolvedShape(axis_names=("draw",), sizes=(3,))

# Concrete jax.Array — silently round-trips through host memory, losing
# device placement / dtype / PyTree identity.
pv = ParameterValue(value=jnp.asarray([1.0, 2.0, 3.0]), shape=shape)
print(type(pv.value).__module__)  # 'numpy'

# jax.vmap tracer — hard error.
def make(x):
    return ParameterValue(value=jnp.broadcast_to(x, (3,)), shape=shape)
jax.vmap(make)(jnp.asarray([1.0, 2.0, 3.0]))
# jax.errors.TracerArrayConversionError
```

The offending line is `ParameterValue.__post_init__`:

```python
value = np.asarray(self.value, dtype=np.float64)
```

## Brief explanation

`ParameterValue` is a `(value, shape)` record. The shape contract is real
(`ResolvedShape` is checked); the dtype/backend contract — `value:
Float64NDArray` plus the unconditional `np.asarray(..., dtype=np.float64)`
cast — is over-specification. The cast is purely the runtime enforcement
of the static `Float64NDArray` annotation, and it has three real costs:

1. It silently coerces malformed plugin input (`int` arrays, `float32`
   arrays, Python lists) to `float64`, hiding bugs that the static
   type-check already catches.
2. It blocks any non-NumPy array backend from passing through. A concrete
   `jax.Array` survives via host round-trip (correct, but loses device
   placement and zero-copy semantics); a `jax.vmap` / `jit` tracer raises
   `TracerArrayConversionError` outright.
3. It implicitly couples `flepimop2` to the NumPy backend at a layer
   (`ParameterABC`'s return contract) where backend-agnosticism would
   otherwise be free.

The natural fix is the [Python Array API standard][1]: every modern array
library (NumPy ≥ 2.0, JAX ≥ 0.4.32, PyTorch ≥ 2.1, CuPy, dask) implements
`__array_namespace__`. A small `Protocol` in `flepimop2.typing` captures
the contract `ParameterValue` actually needs (`.shape` + array-API
membership) without taking a JAX/PyTorch/etc. dependency:

```python
# flepimop2/typing.py
from typing import Protocol, runtime_checkable

@runtime_checkable
class Array(Protocol):
    """Minimal Array-API duck type used by ``ParameterValue``."""
    @property
    def shape(self) -> tuple[int, ...]: ...
    @property
    def dtype(self) -> object: ...
    def __array_namespace__(
        self, *, api_version: str | None = None
    ) -> object: ...
```

```python
# flepimop2/parameter/abc/__init__.py
@dataclass(frozen=True, slots=True)
class ParameterValue:
    value: Array
    shape: ResolvedShape

    def __post_init__(self) -> None:
        if tuple(self.value.shape) != self.shape.sizes:
            raise ValueError(...)
```

That's the entire change. Producers continue to hand in
`numpy.ndarray` (which satisfies `Array`); JAX-driven producers hand in
`jax.Array` (which also satisfies `Array`); the engine's downstream
`jnp.asarray(...)` is unchanged. Plugins remain responsible for handing
in the correct dtype — which `mypy` already enforces.

### Backwards compatibility

- All existing `ParameterValue(np.asarray(...))` callers in
  `src/flepimop2/_cli/_simulate_command.py`,
  `src/flepimop2/parameter/fixed/__init__.py`, and the test suite are
  unaffected: NumPy arrays satisfy `Array`.
- The shape check is preserved verbatim.
- No third-party dependency added (the `Protocol` is duck-typed; nothing
  imports `jax`, etc.).

### Concrete downstream motivation

In ACCIDDA/flepimop2-demo we have a `numpyro_predictive` process plugin
that draws prior samples via `numpyro.infer.Predictive` and threads them
into a parametric expression via a `ContextVar` substitution. The demo's
plugins are already dual-path (numpy ↔ `jax.numpy`); the only place the
JAX path dies is the `ParameterValue` boundary. With this fix:

- `numpyro_predictive` becomes `jax.vmap(run_one)(prior_samples)` instead
  of N serial JIT'd calls.
- A future `numpyro_inference` plugin (NUTS / SVI) can wrap the engine in
  a single `numpyro` model with an observed-data likelihood — the
  gradient/tracer flow is currently blocked at this exact line.

Happy to send the PR (small — typing additions + the `__post_init__`
trim + a couple of unit tests).

[1]: https://data-apis.org/array-api/latest/

### PR DESCRIPTION
**Pull request by jc-macdonald**
## Description

Replaces the over-narrow `Float64NDArray` annotation on `ParameterValue.value` with a new `flepimop2.typing.Array` `Protocol` capturing the minimal [Python Array API](https://data-apis.org/array-api/latest/) surface (`shape`, `dtype`, `__array_namespace__`, `item`). The corresponding `np.asarray(..., dtype=np.float64)` cast in `__post_init__` is removed; only the resolved-shape check remains.

Producers (e.g. `_simulate_command`, `FixedParameter`) already pass NumPy and remain unchanged. NumPy-only engine consumers (the in-tree dummy engine, the `solve_ivp` quickstart, and the integration `euler` engine) now `np.asarray(...)` explicitly at the boundary, which is the correct place to assert the engine's backend requirement.

This unblocks engine-level `jax.vmap` / `jit` / numpyro inference workflows that previously hit `TracerArrayConversionError` on the implicit cast.

### What changed

- `src/flepimop2/typing.py`: Add `@runtime_checkable class Array(Protocol)` and export it.
- `src/flepimop2/parameter/abc/__init__.py`: Field `value: Array` (was `Float64NDArray`); `__post_init__` only validates shape.
- `tests/parameter/test_parameter_value_class.py`: New tests — NumPy round-trip is identity (no copy), shape mismatch raises, JAX `Array` accepted verbatim, `jax.vmap` over a `ParameterValue` factory survives. JAX tests `pytest.importorskip` cleanly when JAX is unavailable.
- `tests/engine/engine_wrapper_assets/dummy_engine.py`, `tests/integration/external_provider/euler.py`, `docs/assets/quickstart_workflow/solve_ivp.py`: Explicit `np.asarray(pv.value)` at the engine boundary (no behaviour change).
- `CHANGELOG.md`: Entry under `Unreleased` → `Fixed`.

## Related issues

Closes #270.

## Checklist

- [x] I have read the [PR process](https://github.com/ACCIDDA/flepimop2/blob/main/CONTRIBUTING.md).
- [x] I have run `just ci` and confirmed all checks pass.
- [x] I have updated the `CHANGELOG.md` with a description of the changes made in this PR.

### CODE REVIEW

**Review by TimothyWillard — CHANGES_REQUESTED** [review]
I think converting `ParameterValue.value: Float64NDArray` to `ParameterValue.value: Array` is a pretty big API change and a downgrade for modules. Ultimately this pushes the responsibility of working with several different array implementations down to the implementors. I think this was going to be inevitable to some degree anyways, but I think maybe we could be a bit more upfront about it by having modules advertise (similar to how `SystemABC` advertises its state change via the `state_change: StateChangeEnum` attribute) and then convert `ParameterValue` to a generic over `value: T`. Furthermore, I think this `Array` protocol is way too loose since numpy string arrays (which cannot be multiplied) satisfy this type.

**Comment by pearsonca** [comment]
> I think converting `ParameterValue.value: Float64NDArray` to `ParameterValue.value: Array` is a pretty big API change and a downgrade for modules. Ultimately this pushes the responsibility of working with several different array implementations down to the implementors. I think this was going to be inevitable to some degree anyways, but I think maybe we could be a bit more upfront about it by having modules advertise (similar to how `SystemABC` advertises its state change via the `state_change: StateChangeEnum` attribute) and then convert `ParameterValue` to a generic over `value: T`. Furthermore, I think this `Array` protocol is way too loose since numpy string arrays (which cannot be multiplied) satisfy this type.

I think there's a bit of work to do here, but that its also conceptually in the right direction.

We need to figure out where 1) determination of need for conversion occurs and 2) if conversion required, where that occurs.

My read of this change is that we want to push developers towards using this more-generic API for arrays. If they can, great; if they cannot, they advertise which particular one they require, and pipeline manages translation.

**Comment by jc-macdonald** [comment]
> I think converting `ParameterValue.value: Float64NDArray` to `ParameterValue.value: Array` is a pretty big API change and a downgrade for modules. Ultimately this pushes the responsibility of working with several different array implementations down to the implementors. I think this was going to be inevitable to some degree anyways, but I think maybe we could be a bit more upfront about it by having modules advertise (similar to how `SystemABC` advertises its state change via the `state_change: StateChangeEnum` attribute) and then convert `ParameterValue` to a generic over `value: T`. Furthermore, I think this `Array` protocol is way too loose since numpy string arrays (which cannot be multiplied) satisfy this type.

I don't think we want to lock into numpy arrays either though because if we force Float64NDArray for example it breaks jax autodiff which is critical for NUTS and other gradient based inference.

**Comment by pearsonca** [comment]
Per slack convo: I agree with a just-do-it approach to this for now, conditioned on introducing a new issue to resolve how we handle being more array-generic.

Particularly, we need introduce test modules to insure that we support developer flexibility, with minimal pain (i.e. they aren't forced to do extensive type conversions), and good performance.

**Comment by jc-macdonald** [comment]
Pushed [`006462b`](https://github.com/ACCIDDA/flepimop2/commit/006462b) addressing @TimothyWillard's feedback on the breadth of the `Array` `Protocol`: `ParameterValue.__post_init__` now rejects non-numeric dtypes (numpy string/object arrays, etc.) by probing `__array_namespace__().isdtype(..., "numeric")` with a `dtype.kind` fallback — no host round-trip on tracer-bearing backends. Two new tests cover string and object dtypes; existing 4 tests still pass.

Filed #275 as the tracking issue for the larger generic-over-`ArrayT` refactor + cross-backend integration test matrix that we agreed should follow this PR to address @pearsonca's request for a tracking issue.

**Review by TimothyWillard — APPROVED** [review]
